src/cargo/core/compiler/context/mod.rs

```diff
@@ -287,10 +287,20 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
         let dest = self.bcx.profiles.get_dir_name();
         let host_layout = Layout::new(self.bcx.ws, None, &dest)?;
         let mut targets = HashMap::new();
-        for kind in self.bcx.build_config.requested_kinds.iter() {
-            if let CompileKind::Target(target) = *kind {
-                let layout = Layout::new(self.bcx.ws, Some(target), &dest)?;
-                targets.insert(target, layout);
+        // Create layouts for each target
+        let all_kinds = self
+            .bcx
+            .build_config
+            .requested_kinds
+            .iter()
+            .copied()
+            .chain(self.bcx.unit_graph.keys().map(|u| u.kind));
```

Comment:
Ah yeah after reading this again, I think that `bcx` will definitely want to have a list (deduped) of all kinds in play for any given compilation.